// @mistralai/mistralai/src/models/components/deploymentlogrecord.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b516079d125e
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type DeploymentLogRecord = {
  timestamp: Date;
  traceId: string;
  spanId: string;
  severityText: string;
  body: string;
  logAttributes: { [k: string]: string };
};

/** @internal */
export const DeploymentLogRecord$inboundSchema: z.ZodType<
  DeploymentLogRecord,
  unknown
> = z.object({
  timestamp: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
  trace_id: z.string(),
  span_id: z.string(),
  severity_text: z.string(),
  body: z.string(),
  log_attributes: z.record(z.string(), z.string()),
}).transform((v) => {
  return remap$(v, {
    "trace_id": "traceId",
    "span_id": "spanId",
    "severity_text": "severityText",
    "log_attributes": "logAttributes",
  });
});

export function deploymentLogRecordFromJSON(
  jsonString: string,
): SafeParseResult<DeploymentLogRecord, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => DeploymentLogRecord$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'DeploymentLogRecord' from JSON`,
  );
}
