// @mistralai/mistralai/src/models/components/deploymentlogsearchresponse.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e6a876334081
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  DeploymentLogRecord,
  DeploymentLogRecord$inboundSchema,
} from "./deploymentlogrecord.js";

export type DeploymentLogSearchResponse = {
  results: Array<DeploymentLogRecord>;
  nextCursor?: string | null | undefined;
};

/** @internal */
export const DeploymentLogSearchResponse$inboundSchema: z.ZodType<
  DeploymentLogSearchResponse,
  unknown
> = z.object({
  results: z.array(DeploymentLogRecord$inboundSchema),
  next_cursor: z.nullable(z.string()).optional(),
}).transform((v) => {
  return remap$(v, {
    "next_cursor": "nextCursor",
  });
});

export function deploymentLogSearchResponseFromJSON(
  jsonString: string,
): SafeParseResult<DeploymentLogSearchResponse, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => DeploymentLogSearchResponse$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'DeploymentLogSearchResponse' from JSON`,
  );
}
