// @mistralai/mistralai/src/models/components/executiontool.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2fd3ed67cc62
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  ExecutionConfig,
  ExecutionConfig$inboundSchema,
} from "./executionconfig.js";

export type ExecutionTool = {
  name: string;
  integrationId: string;
  executionConfig: ExecutionConfig | null;
};

/** @internal */
export const ExecutionTool$inboundSchema: z.ZodType<ExecutionTool, unknown> = z
  .object({
    name: z.string(),
    integration_id: z.string(),
    execution_config: z.nullable(ExecutionConfig$inboundSchema),
  }).transform((v) => {
    return remap$(v, {
      "integration_id": "integrationId",
      "execution_config": "executionConfig",
    });
  });

export function executionToolFromJSON(
  jsonString: string,
): SafeParseResult<ExecutionTool, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ExecutionTool$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ExecutionTool' from JSON`,
  );
}
