// @mistralai/mistralai/src/models/components/executiontraceinforesponse.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b79dab4b50dd
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type ExecutionTraceInfoResponse = {
  /**
   * The ID of the trace, if available
   */
  otelTraceId?: string | null | undefined;
  /**
   * Whether trace data is available in the trace backend for this execution
   */
  hasTraceData: boolean;
};

/** @internal */
export const ExecutionTraceInfoResponse$inboundSchema: z.ZodType<
  ExecutionTraceInfoResponse,
  unknown
> = z.object({
  otel_trace_id: z.nullable(z.string()).optional(),
  has_trace_data: z.boolean().default(false),
}).transform((v) => {
  return remap$(v, {
    "otel_trace_id": "otelTraceId",
    "has_trace_data": "hasTraceData",
  });
});

export function executionTraceInfoResponseFromJSON(
  jsonString: string,
): SafeParseResult<ExecutionTraceInfoResponse, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ExecutionTraceInfoResponse$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ExecutionTraceInfoResponse' from JSON`,
  );
}
