// @mistralai/mistralai/src/models/components/feedresultgetspan.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2522f0b375b4
 */

import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import { GetSpan, GetSpan$inboundSchema } from "./getspan.js";

export type FeedResultGetSpan = {
  results?: Array<GetSpan> | undefined;
  next?: string | null | undefined;
  cursor?: string | null | undefined;
};

/** @internal */
export const FeedResultGetSpan$inboundSchema: z.ZodType<
  FeedResultGetSpan,
  unknown
> = z.object({
  results: z.array(GetSpan$inboundSchema).optional(),
  next: z.nullable(z.string()).optional(),
  cursor: z.nullable(z.string()).optional(),
});

export function feedResultGetSpanFromJSON(
  jsonString: string,
): SafeParseResult<FeedResultGetSpan, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => FeedResultGetSpan$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'FeedResultGetSpan' from JSON`,
  );
}
