// @mistralai/mistralai/src/models/components/feedresultgetspanevaluation.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6bdfbb47b457
 */

import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  GetSpanEvaluation,
  GetSpanEvaluation$inboundSchema,
} from "./getspanevaluation.js";

export type FeedResultGetSpanEvaluation = {
  results?: Array<GetSpanEvaluation> | undefined;
  next?: string | null | undefined;
  cursor?: string | null | undefined;
};

/** @internal */
export const FeedResultGetSpanEvaluation$inboundSchema: z.ZodType<
  FeedResultGetSpanEvaluation,
  unknown
> = z.object({
  results: z.array(GetSpanEvaluation$inboundSchema).optional(),
  next: z.nullable(z.string()).optional(),
  cursor: z.nullable(z.string()).optional(),
});

export function feedResultGetSpanEvaluationFromJSON(
  jsonString: string,
): SafeParseResult<FeedResultGetSpanEvaluation, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => FeedResultGetSpanEvaluation$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'FeedResultGetSpanEvaluation' from JSON`,
  );
}
