// @mistralai/mistralai/src/models/components/getlogfields.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0376653bca71
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  OtelFieldDefinition,
  OtelFieldDefinition$inboundSchema,
} from "./otelfielddefinition.js";

export type GetLogFields = {
  fieldDefinitions: Array<OtelFieldDefinition>;
};

/** @internal */
export const GetLogFields$inboundSchema: z.ZodType<GetLogFields, unknown> = z
  .object({
    field_definitions: z.array(OtelFieldDefinition$inboundSchema),
  }).transform((v) => {
    return remap$(v, {
      "field_definitions": "fieldDefinitions",
    });
  });

export function getLogFieldsFromJSON(
  jsonString: string,
): SafeParseResult<GetLogFields, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => GetLogFields$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'GetLogFields' from JSON`,
  );
}
