// @mistralai/mistralai/src/models/components/ocrcaptionblock.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 839a49096beb
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type OCRCaptionBlock = {
  topLeftX: number;
  topLeftY: number;
  bottomRightX: number;
  bottomRightY: number;
  /**
   * Text/markdown/html content of this block
   */
  content: string;
  type: "caption";
};

/** @internal */
export const OCRCaptionBlock$inboundSchema: z.ZodType<
  OCRCaptionBlock,
  unknown
> = z.object({
  top_left_x: z.int(),
  top_left_y: z.int(),
  bottom_right_x: z.int(),
  bottom_right_y: z.int(),
  content: z.string(),
  type: z.literal("caption"),
}).transform((v) => {
  return remap$(v, {
    "top_left_x": "topLeftX",
    "top_left_y": "topLeftY",
    "bottom_right_x": "bottomRightX",
    "bottom_right_y": "bottomRightY",
  });
});

export function ocrCaptionBlockFromJSON(
  jsonString: string,
): SafeParseResult<OCRCaptionBlock, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => OCRCaptionBlock$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'OCRCaptionBlock' from JSON`,
  );
}
