// @mistralai/mistralai/src/models/components/ocrimageblock.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f3444b64b36e
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type OCRImageBlock = {
  topLeftX: number;
  topLeftY: number;
  bottomRightX: number;
  bottomRightY: number;
  /**
   * Text/markdown/html content of this block
   */
  content: string;
  type: "image";
  /**
   * References the corresponding entry in OCRPageObject.images
   */
  imageId: string;
};

/** @internal */
export const OCRImageBlock$inboundSchema: z.ZodType<OCRImageBlock, unknown> = z
  .object({
    top_left_x: z.int(),
    top_left_y: z.int(),
    bottom_right_x: z.int(),
    bottom_right_y: z.int(),
    content: z.string(),
    type: z.literal("image"),
    image_id: z.string(),
  }).transform((v) => {
    return remap$(v, {
      "top_left_x": "topLeftX",
      "top_left_y": "topLeftY",
      "bottom_right_x": "bottomRightX",
      "bottom_right_y": "bottomRightY",
      "image_id": "imageId",
    });
  });

export function ocrImageBlockFromJSON(
  jsonString: string,
): SafeParseResult<OCRImageBlock, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => OCRImageBlock$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'OCRImageBlock' from JSON`,
  );
}
