// @mistralai/mistralai/src/models/components/publicexecutionconnector.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f9ae4b9b8cb7
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  PublicExecutionConnectionConfig,
  PublicExecutionConnectionConfig$inboundSchema,
} from "./publicexecutionconnectionconfig.js";

export type PublicExecutionConnector = {
  id: string;
  name: string;
  connectionConfig: PublicExecutionConnectionConfig | null;
};

/** @internal */
export const PublicExecutionConnector$inboundSchema: z.ZodType<
  PublicExecutionConnector,
  unknown
> = z.object({
  id: z.string(),
  name: z.string(),
  connection_config: z.nullable(PublicExecutionConnectionConfig$inboundSchema),
}).transform((v) => {
  return remap$(v, {
    "connection_config": "connectionConfig",
  });
});

export function publicExecutionConnectorFromJSON(
  jsonString: string,
): SafeParseResult<PublicExecutionConnector, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => PublicExecutionConnector$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'PublicExecutionConnector' from JSON`,
  );
}
