// @mistralai/mistralai/src/models/components/publicconnectorexecutiondata.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: cf0050336860
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import { ExecutionTool, ExecutionTool$inboundSchema } from "./executiontool.js";
import {
  PublicExecutionConnector,
  PublicExecutionConnector$inboundSchema,
} from "./publicexecutionconnector.js";

export type PublicConnectorExecutionData = {
  integrations: Array<PublicExecutionConnector>;
  tools: Array<ExecutionTool>;
  useConnectorsGateway: boolean;
};

/** @internal */
export const PublicConnectorExecutionData$inboundSchema: z.ZodType<
  PublicConnectorExecutionData,
  unknown
> = z.object({
  integrations: z.array(PublicExecutionConnector$inboundSchema),
  tools: z.array(ExecutionTool$inboundSchema),
  use_connectors_gateway: z.boolean().default(false),
}).transform((v) => {
  return remap$(v, {
    "use_connectors_gateway": "useConnectorsGateway",
  });
});

export function publicConnectorExecutionDataFromJSON(
  jsonString: string,
): SafeParseResult<PublicConnectorExecutionData, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => PublicConnectorExecutionData$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'PublicConnectorExecutionData' from JSON`,
  );
}
