// @mistralai/mistralai/src/models/components/publicexecutionenv.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 56ee1f4abe3b
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  PublicConnectorExecutionData,
  PublicConnectorExecutionData$inboundSchema,
} from "./publicconnectorexecutiondata.js";
import { Tool, Tool$inboundSchema } from "./tool.js";

/**
 * Credentials-free projection of ExecutionEnv for the public /connectors/mistral response.
 */
export type PublicExecutionEnv = {
  tools: Array<Tool>;
  toolExecutionData: PublicConnectorExecutionData;
  errors: Array<string>;
};

/** @internal */
export const PublicExecutionEnv$inboundSchema: z.ZodType<
  PublicExecutionEnv,
  unknown
> = z.object({
  tools: z.array(Tool$inboundSchema),
  tool_execution_data: PublicConnectorExecutionData$inboundSchema,
  errors: z.array(z.string()),
}).transform((v) => {
  return remap$(v, {
    "tool_execution_data": "toolExecutionData",
  });
});

export function publicExecutionEnvFromJSON(
  jsonString: string,
): SafeParseResult<PublicExecutionEnv, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => PublicExecutionEnv$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'PublicExecutionEnv' from JSON`,
  );
}
