// @mistralai/mistralai/src/models/components/registersearchindexrequestvespaschemafield.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 19f7b5b51a50
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import {
  SchemaFieldDataType,
  SchemaFieldDataType$outboundSchema,
} from "./schemafielddatatype.js";
import {
  SchemaFieldIndex,
  SchemaFieldIndex$outboundSchema,
} from "./schemafieldindex.js";
import {
  SchemaFieldRankingType,
  SchemaFieldRankingType$outboundSchema,
} from "./schemafieldrankingtype.js";
import {
  SchemaFieldStorage,
  SchemaFieldStorage$outboundSchema,
} from "./schemafieldstorage.js";

export type RegisterSearchIndexRequestVespaSchemaField = {
  name: string;
  type: SchemaFieldDataType;
  storage: SchemaFieldStorage;
  ranking: SchemaFieldRankingType;
  indexType: SchemaFieldIndex | null;
  multidimensional: boolean;
};

/** @internal */
export type RegisterSearchIndexRequestVespaSchemaField$Outbound = {
  name: string;
  type: string;
  storage: string;
  ranking: string;
  index_type: string | null;
  multidimensional: boolean;
};

/** @internal */
export const RegisterSearchIndexRequestVespaSchemaField$outboundSchema:
  z.ZodType<
    RegisterSearchIndexRequestVespaSchemaField$Outbound,
    RegisterSearchIndexRequestVespaSchemaField
  > = z.object({
    name: z.string(),
    type: SchemaFieldDataType$outboundSchema,
    storage: SchemaFieldStorage$outboundSchema,
    ranking: SchemaFieldRankingType$outboundSchema,
    indexType: z.nullable(SchemaFieldIndex$outboundSchema),
    multidimensional: z.boolean(),
  }).transform((v) => {
    return remap$(v, {
      indexType: "index_type",
    });
  });

export function registerSearchIndexRequestVespaSchemaFieldToJSON(
  registerSearchIndexRequestVespaSchemaField:
    RegisterSearchIndexRequestVespaSchemaField,
): string {
  return JSON.stringify(
    RegisterSearchIndexRequestVespaSchemaField$outboundSchema.parse(
      registerSearchIndexRequestVespaSchemaField,
    ),
  );
}
