// @mistralai/mistralai/src/models/components/registersearchindexrequestvespaindex.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6b775cd19be6
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import {
  RegisterSearchIndexRequestVespaSchemaModel,
  RegisterSearchIndexRequestVespaSchemaModel$Outbound,
  RegisterSearchIndexRequestVespaSchemaModel$outboundSchema,
} from "./registersearchindexrequestvespaschemamodel.js";

export type RegisterSearchIndexRequestVespaIndex = {
  type: "vespa";
  k8sCluster: string;
  k8sNamespace: string;
  vespaInstanceName: string;
  vespaVersion: string;
  schemas: Array<RegisterSearchIndexRequestVespaSchemaModel>;
  queryUrl: string;
};

/** @internal */
export type RegisterSearchIndexRequestVespaIndex$Outbound = {
  type: "vespa";
  k8s_cluster: string;
  k8s_namespace: string;
  vespa_instance_name: string;
  vespa_version: string;
  schemas: Array<RegisterSearchIndexRequestVespaSchemaModel$Outbound>;
  query_url: string;
};

/** @internal */
export const RegisterSearchIndexRequestVespaIndex$outboundSchema: z.ZodType<
  RegisterSearchIndexRequestVespaIndex$Outbound,
  RegisterSearchIndexRequestVespaIndex
> = z.object({
  type: z.literal("vespa"),
  k8sCluster: z.string(),
  k8sNamespace: z.string(),
  vespaInstanceName: z.string(),
  vespaVersion: z.string(),
  schemas: z.array(RegisterSearchIndexRequestVespaSchemaModel$outboundSchema),
  queryUrl: z.string(),
}).transform((v) => {
  return remap$(v, {
    k8sCluster: "k8s_cluster",
    k8sNamespace: "k8s_namespace",
    vespaInstanceName: "vespa_instance_name",
    vespaVersion: "vespa_version",
    queryUrl: "query_url",
  });
});

export function registerSearchIndexRequestVespaIndexToJSON(
  registerSearchIndexRequestVespaIndex: RegisterSearchIndexRequestVespaIndex,
): string {
  return JSON.stringify(
    RegisterSearchIndexRequestVespaIndex$outboundSchema.parse(
      registerSearchIndexRequestVespaIndex,
    ),
  );
}
