// @mistralai/mistralai/src/models/components/registersearchindexrequestvespaschemamodel.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e0bc4fa950d9
 */

import * as z from "zod/v4";
import {
  RegisterSearchIndexRequestVespaSchemaField,
  RegisterSearchIndexRequestVespaSchemaField$Outbound,
  RegisterSearchIndexRequestVespaSchemaField$outboundSchema,
} from "./registersearchindexrequestvespaschemafield.js";

export type RegisterSearchIndexRequestVespaSchemaModel = {
  name: string;
  fields: Array<RegisterSearchIndexRequestVespaSchemaField>;
  sd: string;
};

/** @internal */
export type RegisterSearchIndexRequestVespaSchemaModel$Outbound = {
  name: string;
  fields: Array<RegisterSearchIndexRequestVespaSchemaField$Outbound>;
  sd: string;
};

/** @internal */
export const RegisterSearchIndexRequestVespaSchemaModel$outboundSchema:
  z.ZodType<
    RegisterSearchIndexRequestVespaSchemaModel$Outbound,
    RegisterSearchIndexRequestVespaSchemaModel
  > = z.object({
    name: z.string(),
    fields: z.array(RegisterSearchIndexRequestVespaSchemaField$outboundSchema),
    sd: z.string(),
  });

export function registerSearchIndexRequestVespaSchemaModelToJSON(
  registerSearchIndexRequestVespaSchemaModel:
    RegisterSearchIndexRequestVespaSchemaModel,
): string {
  return JSON.stringify(
    RegisterSearchIndexRequestVespaSchemaModel$outboundSchema.parse(
      registerSearchIndexRequestVespaSchemaModel,
    ),
  );
}
