// @mistralai/mistralai/src/models/components/registersearchindexrequestindex.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3fbd3832dda8
 */

import * as z from "zod/v4";
import { ClosedEnum } from "../../types/enums.js";
import {
  RegisterSearchIndexRequestVespaIndex,
  RegisterSearchIndexRequestVespaIndex$Outbound,
  RegisterSearchIndexRequestVespaIndex$outboundSchema,
} from "./registersearchindexrequestvespaindex.js";

export const RegisterSearchIndexRequestIndexStatus = {
  Online: "online",
  Offline: "offline",
} as const;
export type RegisterSearchIndexRequestIndexStatus = ClosedEnum<
  typeof RegisterSearchIndexRequestIndexStatus
>;

export type RegisterSearchIndexRequestIndexIndex =
  RegisterSearchIndexRequestVespaIndex;

export type RegisterSearchIndexRequestIndex = {
  name: string;
  status?: RegisterSearchIndexRequestIndexStatus | undefined;
  index: RegisterSearchIndexRequestVespaIndex;
};

/** @internal */
export const RegisterSearchIndexRequestIndexStatus$outboundSchema: z.ZodEnum<
  typeof RegisterSearchIndexRequestIndexStatus
> = z.enum(RegisterSearchIndexRequestIndexStatus);

/** @internal */
export type RegisterSearchIndexRequestIndexIndex$Outbound =
  RegisterSearchIndexRequestVespaIndex$Outbound;

/** @internal */
export const RegisterSearchIndexRequestIndexIndex$outboundSchema: z.ZodType<
  RegisterSearchIndexRequestIndexIndex$Outbound,
  RegisterSearchIndexRequestIndexIndex
> = RegisterSearchIndexRequestVespaIndex$outboundSchema;

export function registerSearchIndexRequestIndexIndexToJSON(
  registerSearchIndexRequestIndexIndex: RegisterSearchIndexRequestIndexIndex,
): string {
  return JSON.stringify(
    RegisterSearchIndexRequestIndexIndex$outboundSchema.parse(
      registerSearchIndexRequestIndexIndex,
    ),
  );
}

/** @internal */
export type RegisterSearchIndexRequestIndex$Outbound = {
  name: string;
  status: string;
  index: RegisterSearchIndexRequestVespaIndex$Outbound;
};

/** @internal */
export const RegisterSearchIndexRequestIndex$outboundSchema: z.ZodType<
  RegisterSearchIndexRequestIndex$Outbound,
  RegisterSearchIndexRequestIndex
> = z.object({
  name: z.string(),
  status: RegisterSearchIndexRequestIndexStatus$outboundSchema.default(
    "offline",
  ),
  index: RegisterSearchIndexRequestVespaIndex$outboundSchema,
});

export function registerSearchIndexRequestIndexToJSON(
  registerSearchIndexRequestIndex: RegisterSearchIndexRequestIndex,
): string {
  return JSON.stringify(
    RegisterSearchIndexRequestIndex$outboundSchema.parse(
      registerSearchIndexRequestIndex,
    ),
  );
}
