// @mistralai/mistralai/src/models/components/updateindexmetricsrequestindexmetricsonline.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a9e48e746ac8
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import {
  UpdateIndexMetricsRequestSchemaMetrics,
  UpdateIndexMetricsRequestSchemaMetrics$Outbound,
  UpdateIndexMetricsRequestSchemaMetrics$outboundSchema,
} from "./updateindexmetricsrequestschemametrics.js";

export type UpdateIndexMetricsRequestIndexMetricsOnline = {
  status: "online";
  documentCount: number;
  schemaMetrics: Array<UpdateIndexMetricsRequestSchemaMetrics>;
};

/** @internal */
export type UpdateIndexMetricsRequestIndexMetricsOnline$Outbound = {
  status: "online";
  document_count: number;
  schema_metrics: Array<UpdateIndexMetricsRequestSchemaMetrics$Outbound>;
};

/** @internal */
export const UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema:
  z.ZodType<
    UpdateIndexMetricsRequestIndexMetricsOnline$Outbound,
    UpdateIndexMetricsRequestIndexMetricsOnline
  > = z.object({
    status: z.literal("online"),
    documentCount: z.int(),
    schemaMetrics: z.array(
      UpdateIndexMetricsRequestSchemaMetrics$outboundSchema,
    ),
  }).transform((v) => {
    return remap$(v, {
      documentCount: "document_count",
      schemaMetrics: "schema_metrics",
    });
  });

export function updateIndexMetricsRequestIndexMetricsOnlineToJSON(
  updateIndexMetricsRequestIndexMetricsOnline:
    UpdateIndexMetricsRequestIndexMetricsOnline,
): string {
  return JSON.stringify(
    UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema.parse(
      updateIndexMetricsRequestIndexMetricsOnline,
    ),
  );
}
