// @mistralai/mistralai/src/models/components/updateindexmetricsrequestindexmetricsoffline.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0b243b44976c
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type UpdateIndexMetricsRequestIndexMetricsOffline = {
  status: "offline";
  clearMetrics?: boolean | undefined;
};

/** @internal */
export type UpdateIndexMetricsRequestIndexMetricsOffline$Outbound = {
  status: "offline";
  clear_metrics: boolean;
};

/** @internal */
export const UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema:
  z.ZodType<
    UpdateIndexMetricsRequestIndexMetricsOffline$Outbound,
    UpdateIndexMetricsRequestIndexMetricsOffline
  > = z.object({
    status: z.literal("offline"),
    clearMetrics: z.boolean().default(false),
  }).transform((v) => {
    return remap$(v, {
      clearMetrics: "clear_metrics",
    });
  });

export function updateIndexMetricsRequestIndexMetricsOfflineToJSON(
  updateIndexMetricsRequestIndexMetricsOffline:
    UpdateIndexMetricsRequestIndexMetricsOffline,
): string {
  return JSON.stringify(
    UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema.parse(
      updateIndexMetricsRequestIndexMetricsOffline,
    ),
  );
}
