// @mistralai/mistralai/src/models/components/updateindexmetricsrequestschemametrics.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a784723f090a
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type UpdateIndexMetricsRequestSchemaMetrics = {
  name: string;
  documentCount: number;
};

/** @internal */
export type UpdateIndexMetricsRequestSchemaMetrics$Outbound = {
  name: string;
  document_count: number;
};

/** @internal */
export const UpdateIndexMetricsRequestSchemaMetrics$outboundSchema: z.ZodType<
  UpdateIndexMetricsRequestSchemaMetrics$Outbound,
  UpdateIndexMetricsRequestSchemaMetrics
> = z.object({
  name: z.string(),
  documentCount: z.int(),
}).transform((v) => {
  return remap$(v, {
    documentCount: "document_count",
  });
});

export function updateIndexMetricsRequestSchemaMetricsToJSON(
  updateIndexMetricsRequestSchemaMetrics:
    UpdateIndexMetricsRequestSchemaMetrics,
): string {
  return JSON.stringify(
    UpdateIndexMetricsRequestSchemaMetrics$outboundSchema.parse(
      updateIndexMetricsRequestSchemaMetrics,
    ),
  );
}
