// @mistralai/mistralai/src/models/components/vespagetretrievableresponseretrievable.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 57f152cfd08b
 */

import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type VespaGetRetrievableResponseRetrievable = {
  id: string;
  fields: { [k: string]: any };
};

/** @internal */
export const VespaGetRetrievableResponseRetrievable$inboundSchema: z.ZodType<
  VespaGetRetrievableResponseRetrievable,
  unknown
> = z.object({
  id: z.string(),
  fields: z.record(z.string(), z.any()),
});

export function vespaGetRetrievableResponseRetrievableFromJSON(
  jsonString: string,
): SafeParseResult<VespaGetRetrievableResponseRetrievable, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) =>
      VespaGetRetrievableResponseRetrievable$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'VespaGetRetrievableResponseRetrievable' from JSON`,
  );
}
