// @mistralai/mistralai/src/models/components/workflowscheduletriggerrequest.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d5788fe8e7b0
 */

import * as z from "zod/v4";
import {
  ScheduleOverlapPolicy,
  ScheduleOverlapPolicy$outboundSchema,
} from "./scheduleoverlappolicy.js";

export type WorkflowScheduleTriggerRequest = {
  /**
   * Optional overlap policy override to use for the immediate trigger.
   */
  overlap?: ScheduleOverlapPolicy | null | undefined;
};

/** @internal */
export type WorkflowScheduleTriggerRequest$Outbound = {
  overlap?: number | null | undefined;
};

/** @internal */
export const WorkflowScheduleTriggerRequest$outboundSchema: z.ZodType<
  WorkflowScheduleTriggerRequest$Outbound,
  WorkflowScheduleTriggerRequest
> = z.object({
  overlap: z.nullable(ScheduleOverlapPolicy$outboundSchema).optional(),
});

export function workflowScheduleTriggerRequestToJSON(
  workflowScheduleTriggerRequest: WorkflowScheduleTriggerRequest,
): string {
  return JSON.stringify(
    WorkflowScheduleTriggerRequest$outboundSchema.parse(
      workflowScheduleTriggerRequest,
    ),
  );
}
