// @mistralai/mistralai/esm/models/operations/connectorgetauthurlv1.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e84ab59852d5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const ConnectorGetAuthUrlV1Request$outboundSchema = z.object({
    connectorIdOrName: z.string(),
    appReturnUrl: z.nullable(z.string()).optional(),
    methodType: components.OutboundAuthenticationType$outboundSchema.optional(),
    credentialsName: z.nullable(z.string()).optional(),
    githubInstallationLink: z.boolean().default(false),
}).transform((v) => {
    return remap$(v, {
        connectorIdOrName: "connector_id_or_name",
        appReturnUrl: "app_return_url",
        methodType: "method_type",
        credentialsName: "credentials_name",
        githubInstallationLink: "github_installation_link",
    });
});
export function connectorGetAuthUrlV1RequestToJSON(connectorGetAuthUrlV1Request) {
    return JSON.stringify(ConnectorGetAuthUrlV1Request$outboundSchema.parse(connectorGetAuthUrlV1Request));
}
