// @mistralai/mistralai/esm/models/operations/getdeploymentlogs.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: aacf6aa7e767
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/**
 * First-page sort order: 'asc' (oldest first) or 'desc'. Ignored when `cursor` is set.
 */
export const GetDeploymentLogsOrder = {
    Asc: "asc",
    Desc: "desc",
};
/** @internal */
export const GetDeploymentLogsOrder$outboundSchema = z.enum(GetDeploymentLogsOrder);
/** @internal */
export const GetDeploymentLogsRequest$outboundSchema = z.object({
    name: z.string(),
    workerName: z.nullable(z.string()).optional(),
    workflowName: z.nullable(z.string()).optional(),
    after: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    before: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    order: GetDeploymentLogsOrder$outboundSchema.default("asc"),
    cursor: z.nullable(z.string()).optional(),
    limit: z.int().default(50),
}).transform((v) => {
    return remap$(v, {
        workerName: "worker_name",
        workflowName: "workflow_name",
    });
});
export function getDeploymentLogsRequestToJSON(getDeploymentLogsRequest) {
    return JSON.stringify(GetDeploymentLogsRequest$outboundSchema.parse(getDeploymentLogsRequest));
}
