// @mistralai/mistralai/esm/models/operations/getworkflowsv1workflowsget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e8470b4d34a8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import * as components from "../components/index.js";
/**
 * Filter by deployment activity. active=only active, inactive=only inactive, None=no filter
 */
export const DeploymentStatus = {
    Active: "active",
    Inactive: "inactive",
};
/**
 * Sort direction
 */
export const GetWorkflowsV1WorkflowsGetOrder = {
    Asc: "asc",
    Desc: "desc",
};
/** @internal */
export const GetWorkflowsV1WorkflowsGetStatus$outboundSchema = smartUnion([
    components.WorkflowExecutionStatus$outboundSchema,
    z.array(components.WorkflowExecutionStatus$outboundSchema),
]);
export function getWorkflowsV1WorkflowsGetStatusToJSON(getWorkflowsV1WorkflowsGetStatus) {
    return JSON.stringify(GetWorkflowsV1WorkflowsGetStatus$outboundSchema.parse(getWorkflowsV1WorkflowsGetStatus));
}
/** @internal */
export const DeploymentStatus$outboundSchema = z.enum(DeploymentStatus);
/** @internal */
export const GetWorkflowsV1WorkflowsGetOrder$outboundSchema = z.enum(GetWorkflowsV1WorkflowsGetOrder);
/** @internal */
export const GetWorkflowsV1WorkflowsGetRequest$outboundSchema = z.object({
    status: z.nullable(smartUnion([
        components.WorkflowExecutionStatus$outboundSchema,
        z.array(components.WorkflowExecutionStatus$outboundSchema),
    ])).optional(),
    includeShared: z.boolean().default(true),
    availableInChatAssistant: z.nullable(z.boolean()).optional(),
    deploymentName: z.nullable(z.array(z.string())).optional(),
    deploymentStatus: z.nullable(DeploymentStatus$outboundSchema).optional(),
    archived: z.nullable(z.boolean()).optional(),
    tags: z.nullable(z.array(z.string())).optional(),
    sortBy: z.nullable(z.literal("display_name")).optional(),
    order: GetWorkflowsV1WorkflowsGetOrder$outboundSchema.default("asc"),
    cursor: z.nullable(z.string()).optional(),
    limit: z.int().default(50),
    activeOnly: z.boolean().default(false),
    search: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        includeShared: "include_shared",
        availableInChatAssistant: "available_in_chat_assistant",
        deploymentName: "deployment_name",
        deploymentStatus: "deployment_status",
        sortBy: "sort_by",
        activeOnly: "active_only",
    });
});
export function getWorkflowsV1WorkflowsGetRequestToJSON(getWorkflowsV1WorkflowsGetRequest) {
    return JSON.stringify(GetWorkflowsV1WorkflowsGetRequest$outboundSchema.parse(getWorkflowsV1WorkflowsGetRequest));
}
/** @internal */
export const GetWorkflowsV1WorkflowsGetResponse$inboundSchema = z.object({
    Result: components.WorkflowListResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function getWorkflowsV1WorkflowsGetResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetWorkflowsV1WorkflowsGetResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetWorkflowsV1WorkflowsGetResponse' from JSON`);
}
