// @mistralai/mistralai/esm/models/operations/getworkflowexecutionlogs.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 735c781e7e09
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/**
 * First-page sort order: 'asc' (oldest first) or 'desc'. Ignored when `cursor` is set.
 */
export const GetWorkflowExecutionLogsOrder = {
    Asc: "asc",
    Desc: "desc",
};
/** @internal */
export const GetWorkflowExecutionLogsOrder$outboundSchema = z.enum(GetWorkflowExecutionLogsOrder);
/** @internal */
export const GetWorkflowExecutionLogsRequest$outboundSchema = z.object({
    executionId: z.string(),
    runId: z.nullable(z.string()).optional(),
    activityId: z.nullable(z.string()).optional(),
    after: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    before: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    order: GetWorkflowExecutionLogsOrder$outboundSchema.default("asc"),
    cursor: z.nullable(z.string()).optional(),
    limit: z.int().default(50),
}).transform((v) => {
    return remap$(v, {
        executionId: "execution_id",
        runId: "run_id",
        activityId: "activity_id",
    });
});
export function getWorkflowExecutionLogsRequestToJSON(getWorkflowExecutionLogsRequest) {
    return JSON.stringify(GetWorkflowExecutionLogsRequest$outboundSchema.parse(getWorkflowExecutionLogsRequest));
}
