// @mistralai/mistralai/esm/models/operations/listrunsv1workflowsrunsget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4e2617d53d6d
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import * as components from "../components/index.js";
/**
 * Field to sort by
 */
export const SortBy = {
    StartTime: "start_time",
    EndTime: "end_time",
};
/**
 * Sort direction
 */
export const ListRunsV1WorkflowsRunsGetOrder = {
    Asc: "asc",
    Desc: "desc",
};
/** @internal */
export const ListRunsV1WorkflowsRunsGetStatus$outboundSchema = smartUnion([
    components.WorkflowExecutionStatus$outboundSchema,
    z.array(components.WorkflowExecutionStatus$outboundSchema),
]);
export function listRunsV1WorkflowsRunsGetStatusToJSON(listRunsV1WorkflowsRunsGetStatus) {
    return JSON.stringify(ListRunsV1WorkflowsRunsGetStatus$outboundSchema.parse(listRunsV1WorkflowsRunsGetStatus));
}
/** @internal */
export const SortBy$outboundSchema = z.enum(SortBy);
/** @internal */
export const ListRunsV1WorkflowsRunsGetOrder$outboundSchema = z.enum(ListRunsV1WorkflowsRunsGetOrder);
/** @internal */
export const ListRunsV1WorkflowsRunsGetRequest$outboundSchema = z.object({
    workflowIdentifier: z.nullable(z.string()).optional(),
    rootExecutionId: z.nullable(z.string()).optional(),
    search: z.nullable(z.string()).optional(),
    status: z.nullable(smartUnion([
        components.WorkflowExecutionStatus$outboundSchema,
        z.array(components.WorkflowExecutionStatus$outboundSchema),
    ])).optional(),
    deploymentName: z.nullable(z.string()).optional(),
    sortBy: z.nullable(SortBy$outboundSchema).optional(),
    order: ListRunsV1WorkflowsRunsGetOrder$outboundSchema.default("desc"),
    startTimeAfter: z.nullable(z.date().transform(v => v.toISOString()))
        .optional(),
    startTimeBefore: z.nullable(z.date().transform(v => v.toISOString()))
        .optional(),
    endTimeAfter: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    endTimeBefore: z.nullable(z.date().transform(v => v.toISOString()))
        .optional(),
    userId: z.nullable(z.string()).optional(),
    includeInternal: z.boolean().default(true),
    pageSize: z.int().default(50),
    nextPageToken: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        workflowIdentifier: "workflow_identifier",
        rootExecutionId: "root_execution_id",
        deploymentName: "deployment_name",
        sortBy: "sort_by",
        startTimeAfter: "start_time_after",
        startTimeBefore: "start_time_before",
        endTimeAfter: "end_time_after",
        endTimeBefore: "end_time_before",
        userId: "user_id",
        includeInternal: "include_internal",
        pageSize: "page_size",
        nextPageToken: "next_page_token",
    });
});
export function listRunsV1WorkflowsRunsGetRequestToJSON(listRunsV1WorkflowsRunsGetRequest) {
    return JSON.stringify(ListRunsV1WorkflowsRunsGetRequest$outboundSchema.parse(listRunsV1WorkflowsRunsGetRequest));
}
/** @internal */
export const ListRunsV1WorkflowsRunsGetResponse$inboundSchema = z.object({
    Result: components.WorkflowExecutionListResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function listRunsV1WorkflowsRunsGetResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListRunsV1WorkflowsRunsGetResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListRunsV1WorkflowsRunsGetResponse' from JSON`);
}
