// @mistralai/mistralai/esm/models/operations/listdeploymentsv1workflowsdeploymentsget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: afb19c640ecc
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/** @internal */
export const ListDeploymentsV1WorkflowsDeploymentsGetRequest$outboundSchema = z.object({
    activeOnly: z.boolean().default(true),
    isHardened: z.nullable(z.boolean()).optional(),
    workflowName: z.nullable(z.string()).optional(),
    search: z.nullable(z.string()).optional(),
    limit: z.nullable(z.int()).optional(),
    cursor: z.nullable(z.string()).optional(),
    workspaceId: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        activeOnly: "active_only",
        isHardened: "is_hardened",
        workflowName: "workflow_name",
        workspaceId: "workspace_id",
    });
});
export function listDeploymentsV1WorkflowsDeploymentsGetRequestToJSON(listDeploymentsV1WorkflowsDeploymentsGetRequest) {
    return JSON.stringify(ListDeploymentsV1WorkflowsDeploymentsGetRequest$outboundSchema.parse(listDeploymentsV1WorkflowsDeploymentsGetRequest));
}
