// @mistralai/mistralai/esm/models/operations/searchspanevaluationsv1observabilityspansevaluationssearchpost.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5dd192d44063
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$outboundSchema = z.object({
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    pageSize: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
    spanEvaluationsRequest: components.SpanEvaluationsRequest$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        spanEvaluationsRequest: "SpanEvaluationsRequest",
    });
});
export function searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequestToJSON(searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest) {
    return JSON.stringify(SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$outboundSchema
        .parse(searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest));
}
