// @mistralai/mistralai/esm/models/operations/searchlogsv1observabilitylogssearchpost.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7dcabf6fb49a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const SearchLogsV1ObservabilityLogsSearchPostRequest$outboundSchema = z.object({
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    pageSize: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
    logsRequest: components.LogsRequest$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        logsRequest: "LogsRequest",
    });
});
export function searchLogsV1ObservabilityLogsSearchPostRequestToJSON(searchLogsV1ObservabilityLogsSearchPostRequest) {
    return JSON.stringify(SearchLogsV1ObservabilityLogsSearchPostRequest$outboundSchema.parse(searchLogsV1ObservabilityLogsSearchPostRequest));
}
