// @mistralai/mistralai/esm/models/operations/searchlatestspanevaluationsv1observabilityspansevaluationssearchlatestpost.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f09578ad785c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const SearchLatestSpanEvaluationsV1ObservabilitySpansEvaluationsSearchLatestPostRequest$outboundSchema = z.object({
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    pageSize: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
    spanEvaluationsRequest: components.SpanEvaluationsRequest$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        spanEvaluationsRequest: "SpanEvaluationsRequest",
    });
});
export function searchLatestSpanEvaluationsV1ObservabilitySpansEvaluationsSearchLatestPostRequestToJSON(searchLatestSpanEvaluationsV1ObservabilitySpansEvaluationsSearchLatestPostRequest) {
    return JSON.stringify(SearchLatestSpanEvaluationsV1ObservabilitySpansEvaluationsSearchLatestPostRequest$outboundSchema
        .parse(searchLatestSpanEvaluationsV1ObservabilitySpansEvaluationsSearchLatestPostRequest));
}
