// @mistralai/mistralai/esm/models/operations/streamdeploymentlogs.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6c11f6c5701a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { smartUnion } from "../../types/smartUnion.js";
import * as components from "../components/index.js";
export const StreamDeploymentLogsEvent = {
    Log: "log",
    Error: "error",
};
/** @internal */
export const StreamDeploymentLogsRequest$outboundSchema = z.object({
    name: z.string(),
    workerName: z.nullable(z.string()).optional(),
    workflowName: z.nullable(z.string()).optional(),
    after: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    lastEventId: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        workerName: "worker_name",
        workflowName: "workflow_name",
        lastEventId: "last_event_id",
    });
});
export function streamDeploymentLogsRequestToJSON(streamDeploymentLogsRequest) {
    return JSON.stringify(StreamDeploymentLogsRequest$outboundSchema.parse(streamDeploymentLogsRequest));
}
/** @internal */
export const StreamDeploymentLogsEvent$inboundSchema = openEnums.inboundSchema(StreamDeploymentLogsEvent);
/** @internal */
export const StreamDeploymentLogsData$inboundSchema = smartUnion([
    components.DeploymentLogRecord$inboundSchema,
    components.StreamError$inboundSchema,
]);
export function streamDeploymentLogsDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => StreamDeploymentLogsData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'StreamDeploymentLogsData' from JSON`);
}
/** @internal */
export const StreamDeploymentLogsResponseBody$inboundSchema = z.object({
    event: StreamDeploymentLogsEvent$inboundSchema.optional(),
    id: z.string().optional(),
    data: z.string().optional().transform((v, ctx) => {
        if (v === undefined)
            return undefined;
        try {
            return JSON.parse(v);
        }
        catch (err) {
            ctx.addIssue({
                input: v,
                code: "custom",
                message: `malformed json: ${err}`,
            });
            return z.NEVER;
        }
    }).pipe(smartUnion([
        components.DeploymentLogRecord$inboundSchema,
        components.StreamError$inboundSchema,
    ]).optional()),
});
export function streamDeploymentLogsResponseBodyFromJSON(jsonString) {
    return safeParse(jsonString, (x) => StreamDeploymentLogsResponseBody$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'StreamDeploymentLogsResponseBody' from JSON`);
}
