// @mistralai/mistralai/esm/models/operations/updateindexmetricsv1ragindexesindexindexidmetricsput.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0419a3478b8c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const MetricsData$outboundSchema = z.union([
    components.UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema,
    components.UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema,
]);
export function metricsDataToJSON(metricsData) {
    return JSON.stringify(MetricsData$outboundSchema.parse(metricsData));
}
/** @internal */
export const UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$outboundSchema = z.object({
    indexId: z.string(),
    requestBody: z.union([
        components.UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema,
        components.UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema,
    ]),
}).transform((v) => {
    return remap$(v, {
        indexId: "index_id",
        requestBody: "RequestBody",
    });
});
export function updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequestToJSON(updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest) {
    return JSON.stringify(UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$outboundSchema
        .parse(updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest));
}
