// @mistralai/mistralai/esm/sdk/observability.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9c0e070af33d
 */
import { ClientSDK } from "../lib/sdks.js";
import { Campaigns } from "./campaigns.js";
import { ChatCompletionEvents } from "./chatcompletionevents.js";
import { Datasets } from "./datasets.js";
import { Judges } from "./judges.js";
import { Logs } from "./logs.js";
import { Spans } from "./spans.js";
import { Traces } from "./traces.js";
export class Observability extends ClientSDK {
    _chatCompletionEvents;
    get chatCompletionEvents() {
        return (this._chatCompletionEvents ?? (this._chatCompletionEvents = new ChatCompletionEvents(this._options)));
    }
    _judges;
    get judges() {
        return (this._judges ?? (this._judges = new Judges(this._options)));
    }
    _campaigns;
    get campaigns() {
        return (this._campaigns ?? (this._campaigns = new Campaigns(this._options)));
    }
    _datasets;
    get datasets() {
        return (this._datasets ?? (this._datasets = new Datasets(this._options)));
    }
    _logs;
    get logs() {
        return (this._logs ?? (this._logs = new Logs(this._options)));
    }
    _traces;
    get traces() {
        return (this._traces ?? (this._traces = new Traces(this._options)));
    }
    _spans;
    get spans() {
        return (this._spans ?? (this._spans = new Spans(this._options)));
    }
}
