// @mistralai/mistralai/esm/sdk/spans.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c69e203e4392
 */
import { betaObservabilitySpansFetchSpanEvalFieldOptions } from "../funcs/betaObservabilitySpansFetchSpanEvalFieldOptions.js";
import { betaObservabilitySpansFetchSpanFieldOptions } from "../funcs/betaObservabilitySpansFetchSpanFieldOptions.js";
import { betaObservabilitySpansListSpanEvalFields } from "../funcs/betaObservabilitySpansListSpanEvalFields.js";
import { betaObservabilitySpansListSpanFields } from "../funcs/betaObservabilitySpansListSpanFields.js";
import { betaObservabilitySpansSearchLatestSpanEvaluations } from "../funcs/betaObservabilitySpansSearchLatestSpanEvaluations.js";
import { betaObservabilitySpansSearchSpanEvaluations } from "../funcs/betaObservabilitySpansSearchSpanEvaluations.js";
import { betaObservabilitySpansSearchSpans } from "../funcs/betaObservabilitySpansSearchSpans.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Spans extends ClientSDK {
    /**
     * Search spans
     */
    async searchSpans(request, options) {
        return unwrapAsync(betaObservabilitySpansSearchSpans(this, request, options));
    }
    /**
     * Search span evaluations
     */
    async searchSpanEvaluations(request, options) {
        return unwrapAsync(betaObservabilitySpansSearchSpanEvaluations(this, request, options));
    }
    /**
     * Search latest span evaluations
     */
    async searchLatestSpanEvaluations(request, options) {
        return unwrapAsync(betaObservabilitySpansSearchLatestSpanEvaluations(this, request, options));
    }
    /**
     * Get span field definitions
     */
    async listSpanFields(options) {
        return unwrapAsync(betaObservabilitySpansListSpanFields(this, options));
    }
    /**
     * Get span evaluation field definitions
     */
    async listSpanEvalFields(options) {
        return unwrapAsync(betaObservabilitySpansListSpanEvalFields(this, options));
    }
    /**
     * Get options for a span field
     */
    async fetchSpanFieldOptions(request, options) {
        return unwrapAsync(betaObservabilitySpansFetchSpanFieldOptions(this, request, options));
    }
    /**
     * Get options for a span evaluation field
     */
    async fetchSpanEvalFieldOptions(request, options) {
        return unwrapAsync(betaObservabilitySpansFetchSpanEvalFieldOptions(this, request, options));
    }
}
