// @mistralai/mistralai/esm/sdk/traces.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4845e9e6a8d1
 */
import { betaObservabilityTracesFetchOptions } from "../funcs/betaObservabilityTracesFetchOptions.js";
import { betaObservabilityTracesGetSpanById } from "../funcs/betaObservabilityTracesGetSpanById.js";
import { betaObservabilityTracesGetTraceById } from "../funcs/betaObservabilityTracesGetTraceById.js";
import { betaObservabilityTracesGetTraceFields } from "../funcs/betaObservabilityTracesGetTraceFields.js";
import { betaObservabilityTracesGetTraceSpans } from "../funcs/betaObservabilityTracesGetTraceSpans.js";
import { betaObservabilityTracesSearch } from "../funcs/betaObservabilityTracesSearch.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Traces extends ClientSDK {
    /**
     * Search traces
     */
    async search(request, options) {
        return unwrapAsync(betaObservabilityTracesSearch(this, request, options));
    }
    /**
     * Get trace field definitions
     */
    async getTraceFields(options) {
        return unwrapAsync(betaObservabilityTracesGetTraceFields(this, options));
    }
    /**
     * Get trace by id
     */
    async getTraceById(request, options) {
        return unwrapAsync(betaObservabilityTracesGetTraceById(this, request, options));
    }
    /**
     * Get trace spans
     */
    async getTraceSpans(request, options) {
        return unwrapAsync(betaObservabilityTracesGetTraceSpans(this, request, options));
    }
    /**
     * Get options for a trace field
     */
    async fetchOptions(request, options) {
        return unwrapAsync(betaObservabilityTracesFetchOptions(this, request, options));
    }
    /**
     * Get span by id
     */
    async getSpanById(request, options) {
        return unwrapAsync(betaObservabilityTracesGetSpanById(this, request, options));
    }
}
