// @mistralai/mistralai/esm/models/components/createconnectorrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ecc43dadef3c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { AuthData$outboundSchema, } from "./authdata.js";
import { ExtendedOAuthServerMetadata$outboundSchema, } from "./extendedoauthservermetadata.js";
import { ResourceVisibility$outboundSchema, } from "./resourcevisibility.js";
/** @internal */
export const CreateConnectorRequest$outboundSchema = z.object({
    protocol: z.literal("mcp").default("mcp"),
    name: z.string(),
    title: z.nullable(z.string()).optional(),
    description: z.string(),
    iconUrl: z.nullable(z.string()).optional(),
    visibility: ResourceVisibility$outboundSchema.optional(),
    server: z.string(),
    headers: z.nullable(z.record(z.string(), z.any())).optional(),
    authData: z.nullable(AuthData$outboundSchema).optional(),
    oauth2ServerMetadata: z.nullable(ExtendedOAuthServerMetadata$outboundSchema)
        .optional(),
    oauth2ServerMetadataUrl: z.nullable(z.string()).optional(),
    systemPrompt: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        iconUrl: "icon_url",
        authData: "auth_data",
        oauth2ServerMetadata: "oauth2_server_metadata",
        oauth2ServerMetadataUrl: "oauth2_server_metadata_url",
        systemPrompt: "system_prompt",
    });
});
export function createConnectorRequestToJSON(createConnectorRequest) {
    return JSON.stringify(CreateConnectorRequest$outboundSchema.parse(createConnectorRequest));
}
