// @mistralai/mistralai/esm/models/components/getsearchindexsummaryresponseindex.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 49953f07ad83
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { GetSearchIndexSummaryResponseVespaIndex$inboundSchema, } from "./getsearchindexsummaryresponsevespaindex.js";
export const GetSearchIndexSummaryResponseIndexStatus = {
    Online: "online",
    Offline: "offline",
};
/** @internal */
export const GetSearchIndexSummaryResponseIndexStatus$inboundSchema = openEnums.inboundSchema(GetSearchIndexSummaryResponseIndexStatus);
/** @internal */
export const GetSearchIndexSummaryResponseIndexIndex$inboundSchema = GetSearchIndexSummaryResponseVespaIndex$inboundSchema;
export function getSearchIndexSummaryResponseIndexIndexFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetSearchIndexSummaryResponseIndexIndex$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetSearchIndexSummaryResponseIndexIndex' from JSON`);
}
/** @internal */
export const GetSearchIndexSummaryResponseIndex$inboundSchema = z.object({
    id: z.string(),
    name: z.string(),
    creator_id: z.string(),
    document_count: z.int(),
    status: GetSearchIndexSummaryResponseIndexStatus$inboundSchema,
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    modified_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    index: GetSearchIndexSummaryResponseVespaIndex$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "creator_id": "creatorId",
        "document_count": "documentCount",
        "created_at": "createdAt",
        "modified_at": "modifiedAt",
    });
});
export function getSearchIndexSummaryResponseIndexFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetSearchIndexSummaryResponseIndex$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetSearchIndexSummaryResponseIndex' from JSON`);
}
