// @mistralai/mistralai/esm/models/components/getsearchindexschemadetailresponseschemamodel.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3d179cbf1510
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { GetSearchIndexSchemaDetailResponseField$inboundSchema, } from "./getsearchindexschemadetailresponsefield.js";
/** @internal */
export const GetSearchIndexSchemaDetailResponseSchemaModel$inboundSchema = z.object({
    name: z.string(),
    summary: z.nullable(z.string()),
    embedding_dimensions: z.nullable(z.int()),
    fields: z.array(GetSearchIndexSchemaDetailResponseField$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "embedding_dimensions": "embeddingDimensions",
    });
});
export function getSearchIndexSchemaDetailResponseSchemaModelFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetSearchIndexSchemaDetailResponseSchemaModel$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetSearchIndexSchemaDetailResponseSchemaModel' from JSON`);
}
