// @mistralai/mistralai/esm/models/components/getsearchindexschemadetailresponsefield.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: aad31af8795b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { SchemaFieldDataType$inboundSchema, } from "./schemafielddatatype.js";
import { SchemaFieldIndex$inboundSchema, } from "./schemafieldindex.js";
import { SchemaFieldStorage$inboundSchema, } from "./schemafieldstorage.js";
/** @internal */
export const GetSearchIndexSchemaDetailResponseField$inboundSchema = z.object({
    name: z.string(),
    type: SchemaFieldDataType$inboundSchema,
    storage: SchemaFieldStorage$inboundSchema,
    index_type: z.nullable(SchemaFieldIndex$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "index_type": "indexType",
    });
});
export function getSearchIndexSchemaDetailResponseFieldFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetSearchIndexSchemaDetailResponseField$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetSearchIndexSchemaDetailResponseField' from JSON`);
}
