// @mistralai/mistralai/esm/models/components/ocrpageobject.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9f1e92f79055
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import { OCRAsideTextBlock$inboundSchema, } from "./ocrasidetextblock.js";
import { OCRCaptionBlock$inboundSchema, } from "./ocrcaptionblock.js";
import { OCRCodeBlock$inboundSchema } from "./ocrcodeblock.js";
import { OCREquationBlock$inboundSchema, } from "./ocrequationblock.js";
import { OCRFooterBlock$inboundSchema, } from "./ocrfooterblock.js";
import { OCRHeaderBlock$inboundSchema, } from "./ocrheaderblock.js";
import { OCRImageBlock$inboundSchema } from "./ocrimageblock.js";
import { OCRImageObject$inboundSchema, } from "./ocrimageobject.js";
import { OCRListBlock$inboundSchema } from "./ocrlistblock.js";
import { OCRPageConfidenceScores$inboundSchema, } from "./ocrpageconfidencescores.js";
import { OCRPageDimensions$inboundSchema, } from "./ocrpagedimensions.js";
import { OCRReferencesBlock$inboundSchema, } from "./ocrreferencesblock.js";
import { OCRSignatureBlock$inboundSchema, } from "./ocrsignatureblock.js";
import { OCRTableBlock$inboundSchema } from "./ocrtableblock.js";
import { OCRTableObject$inboundSchema, } from "./ocrtableobject.js";
import { OCRTextBlock$inboundSchema } from "./ocrtextblock.js";
import { OCRTitleBlock$inboundSchema } from "./ocrtitleblock.js";
/** @internal */
export const Block$inboundSchema = discriminatedUnion("type", {
    aside_text: OCRAsideTextBlock$inboundSchema,
    caption: OCRCaptionBlock$inboundSchema,
    code: OCRCodeBlock$inboundSchema,
    equation: OCREquationBlock$inboundSchema,
    footer: OCRFooterBlock$inboundSchema,
    header: OCRHeaderBlock$inboundSchema,
    image: OCRImageBlock$inboundSchema,
    list: OCRListBlock$inboundSchema,
    references: OCRReferencesBlock$inboundSchema,
    signature: OCRSignatureBlock$inboundSchema,
    table: OCRTableBlock$inboundSchema,
    text: OCRTextBlock$inboundSchema,
    title: OCRTitleBlock$inboundSchema,
});
export function blockFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Block$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Block' from JSON`);
}
/** @internal */
export const OCRPageObject$inboundSchema = z
    .object({
    index: z.int(),
    markdown: z.string(),
    images: z.array(OCRImageObject$inboundSchema),
    tables: z.array(OCRTableObject$inboundSchema).optional(),
    hyperlinks: z.array(z.string()).optional(),
    header: z.nullable(z.string()).optional(),
    footer: z.nullable(z.string()).optional(),
    dimensions: z.nullable(OCRPageDimensions$inboundSchema),
    confidence_scores: z.nullable(OCRPageConfidenceScores$inboundSchema)
        .optional(),
    blocks: z.nullable(z.array(discriminatedUnion("type", {
        aside_text: OCRAsideTextBlock$inboundSchema,
        caption: OCRCaptionBlock$inboundSchema,
        code: OCRCodeBlock$inboundSchema,
        equation: OCREquationBlock$inboundSchema,
        footer: OCRFooterBlock$inboundSchema,
        header: OCRHeaderBlock$inboundSchema,
        image: OCRImageBlock$inboundSchema,
        list: OCRListBlock$inboundSchema,
        references: OCRReferencesBlock$inboundSchema,
        signature: OCRSignatureBlock$inboundSchema,
        table: OCRTableBlock$inboundSchema,
        text: OCRTextBlock$inboundSchema,
        title: OCRTitleBlock$inboundSchema,
    }))).optional(),
}).transform((v) => {
    return remap$(v, {
        "confidence_scores": "confidenceScores",
    });
});
export function ocrPageObjectFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OCRPageObject$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OCRPageObject' from JSON`);
}
