// @mistralai/mistralai/esm/models/components/otelfielddefinition.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 32bf85982e0b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
export const OtelFieldDefinitionType = {
    Enum: "ENUM",
    Text: "TEXT",
    Int: "INT",
    Float: "FLOAT",
    Bool: "BOOL",
    Timestamp: "TIMESTAMP",
    Array: "ARRAY",
    Map: "MAP",
};
export const OtelFieldDefinitionSupportedOperator = {
    Eq: "eq",
    Neq: "neq",
    Lt: "lt",
    Lte: "lte",
    Gt: "gt",
    Gte: "gte",
    Like: "like",
    Ilike: "ilike",
    NotLike: "not_like",
    NotIlike: "not_ilike",
    Between: "between",
    NotBetween: "not_between",
    In: "in",
    NotIn: "not_in",
    Exists: "exists",
    NotExists: "not_exists",
    Regexp: "regexp",
    NotRegexp: "not_regexp",
    Contains: "contains",
    NotContains: "not_contains",
    Has: "has",
    HasAny: "hasAny",
    HasAll: "hasAll",
    HasToken: "hasToken",
};
/** @internal */
export const OtelFieldDefinitionType$inboundSchema = openEnums.inboundSchema(OtelFieldDefinitionType);
/** @internal */
export const OtelFieldDefinitionSupportedOperator$inboundSchema = openEnums.inboundSchema(OtelFieldDefinitionSupportedOperator);
/** @internal */
export const OtelFieldDefinition$inboundSchema = z.object({
    name: z.string(),
    label: z.string(),
    type: OtelFieldDefinitionType$inboundSchema,
    group: z.nullable(z.string()).optional(),
    supported_operators: z.array(OtelFieldDefinitionSupportedOperator$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "supported_operators": "supportedOperators",
    });
});
export function otelFieldDefinitionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OtelFieldDefinition$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OtelFieldDefinition' from JSON`);
}
