// @mistralai/mistralai/esm/models/components/publicexecutionconnectionconfig.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e8bed813edfc
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ConnectionConfigType$inboundSchema, } from "./connectionconfigtype.js";
import { ToolExecutionConfiguration$inboundSchema, } from "./toolexecutionconfiguration.js";
/** @internal */
export const PublicExecutionConnectionConfig$inboundSchema = z.object({
    type: ConnectionConfigType$inboundSchema,
    server: z.nullable(z.string()).optional(),
    name: z.nullable(z.string()).optional(),
    id: z.nullable(z.string()).optional(),
    tool_configuration: z.nullable(ToolExecutionConfiguration$inboundSchema)
        .optional(),
    hosted_internally: z.boolean().default(false),
}).transform((v) => {
    return remap$(v, {
        "tool_configuration": "toolConfiguration",
        "hosted_internally": "hostedInternally",
    });
});
export function publicExecutionConnectionConfigFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PublicExecutionConnectionConfig$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PublicExecutionConnectionConfig' from JSON`);
}
