// @mistralai/mistralai/esm/models/components/registersearchindexrequestindex.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3fbd3832dda8
 */
import * as z from "zod/v4";
import { RegisterSearchIndexRequestVespaIndex$outboundSchema, } from "./registersearchindexrequestvespaindex.js";
export const RegisterSearchIndexRequestIndexStatus = {
    Online: "online",
    Offline: "offline",
};
/** @internal */
export const RegisterSearchIndexRequestIndexStatus$outboundSchema = z.enum(RegisterSearchIndexRequestIndexStatus);
/** @internal */
export const RegisterSearchIndexRequestIndexIndex$outboundSchema = RegisterSearchIndexRequestVespaIndex$outboundSchema;
export function registerSearchIndexRequestIndexIndexToJSON(registerSearchIndexRequestIndexIndex) {
    return JSON.stringify(RegisterSearchIndexRequestIndexIndex$outboundSchema.parse(registerSearchIndexRequestIndexIndex));
}
/** @internal */
export const RegisterSearchIndexRequestIndex$outboundSchema = z.object({
    name: z.string(),
    status: RegisterSearchIndexRequestIndexStatus$outboundSchema.default("offline"),
    index: RegisterSearchIndexRequestVespaIndex$outboundSchema,
});
export function registerSearchIndexRequestIndexToJSON(registerSearchIndexRequestIndex) {
    return JSON.stringify(RegisterSearchIndexRequestIndex$outboundSchema.parse(registerSearchIndexRequestIndex));
}
