// @mistralai/mistralai/esm/models/components/registersearchindexrequestvespaschemafield.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 19f7b5b51a50
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { SchemaFieldDataType$outboundSchema, } from "./schemafielddatatype.js";
import { SchemaFieldIndex$outboundSchema, } from "./schemafieldindex.js";
import { SchemaFieldRankingType$outboundSchema, } from "./schemafieldrankingtype.js";
import { SchemaFieldStorage$outboundSchema, } from "./schemafieldstorage.js";
/** @internal */
export const RegisterSearchIndexRequestVespaSchemaField$outboundSchema = z.object({
    name: z.string(),
    type: SchemaFieldDataType$outboundSchema,
    storage: SchemaFieldStorage$outboundSchema,
    ranking: SchemaFieldRankingType$outboundSchema,
    indexType: z.nullable(SchemaFieldIndex$outboundSchema),
    multidimensional: z.boolean(),
}).transform((v) => {
    return remap$(v, {
        indexType: "index_type",
    });
});
export function registerSearchIndexRequestVespaSchemaFieldToJSON(registerSearchIndexRequestVespaSchemaField) {
    return JSON.stringify(RegisterSearchIndexRequestVespaSchemaField$outboundSchema.parse(registerSearchIndexRequestVespaSchemaField));
}
