// @mistralai/mistralai/src/models/operations/connectorlistworkspacecredentialsv1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7ba64040b920
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type ConnectorListWorkspaceCredentialsV1Request = {
  connectorIdOrName: string;
  authType?: components.OutboundAuthenticationType | null | undefined;
  fetchDefault?: boolean | undefined;
};

/** @internal */
export type ConnectorListWorkspaceCredentialsV1Request$Outbound = {
  connector_id_or_name: string;
  auth_type?: string | null | undefined;
  fetch_default: boolean;
};

/** @internal */
export const ConnectorListWorkspaceCredentialsV1Request$outboundSchema:
  z.ZodType<
    ConnectorListWorkspaceCredentialsV1Request$Outbound,
    ConnectorListWorkspaceCredentialsV1Request
  > = z.object({
    connectorIdOrName: z.string(),
    authType: z.nullable(components.OutboundAuthenticationType$outboundSchema)
      .optional(),
    fetchDefault: z.boolean().default(false),
  }).transform((v) => {
    return remap$(v, {
      connectorIdOrName: "connector_id_or_name",
      authType: "auth_type",
      fetchDefault: "fetch_default",
    });
  });

export function connectorListWorkspaceCredentialsV1RequestToJSON(
  connectorListWorkspaceCredentialsV1Request:
    ConnectorListWorkspaceCredentialsV1Request,
): string {
  return JSON.stringify(
    ConnectorListWorkspaceCredentialsV1Request$outboundSchema.parse(
      connectorListWorkspaceCredentialsV1Request,
    ),
  );
}
