// @mistralai/mistralai/src/models/operations/connectorlistorganizationcredentialsv1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 92d3577970b6
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type ConnectorListOrganizationCredentialsV1Request = {
  connectorIdOrName: string;
  authType?: components.OutboundAuthenticationType | null | undefined;
  fetchDefault?: boolean | undefined;
};

/** @internal */
export type ConnectorListOrganizationCredentialsV1Request$Outbound = {
  connector_id_or_name: string;
  auth_type?: string | null | undefined;
  fetch_default: boolean;
};

/** @internal */
export const ConnectorListOrganizationCredentialsV1Request$outboundSchema:
  z.ZodType<
    ConnectorListOrganizationCredentialsV1Request$Outbound,
    ConnectorListOrganizationCredentialsV1Request
  > = z.object({
    connectorIdOrName: z.string(),
    authType: z.nullable(components.OutboundAuthenticationType$outboundSchema)
      .optional(),
    fetchDefault: z.boolean().default(false),
  }).transform((v) => {
    return remap$(v, {
      connectorIdOrName: "connector_id_or_name",
      authType: "auth_type",
      fetchDefault: "fetch_default",
    });
  });

export function connectorListOrganizationCredentialsV1RequestToJSON(
  connectorListOrganizationCredentialsV1Request:
    ConnectorListOrganizationCredentialsV1Request,
): string {
  return JSON.stringify(
    ConnectorListOrganizationCredentialsV1Request$outboundSchema.parse(
      connectorListOrganizationCredentialsV1Request,
    ),
  );
}
