// @mistralai/mistralai/src/models/operations/documentlookupv1ragindexesindexindexidschemasschemaschemaidretrievablesretrievabledocumentidget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 22f2ad3a80f1
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest =
  {
    indexId: string;
    schemaId: string;
    /**
     * the native ID in the underlying index
     */
    documentId: string;
  };

/** @internal */
export type DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest$Outbound =
  {
    index_id: string;
    schema_id: string;
    document_id: string;
  };

/** @internal */
export const DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest$outboundSchema:
  z.ZodType<
    DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest$Outbound,
    DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest
  > = z.object({
    indexId: z.string(),
    schemaId: z.string(),
    documentId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      schemaId: "schema_id",
      documentId: "document_id",
    });
  });

export function documentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequestToJSON(
  documentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest:
    DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest,
): string {
  return JSON.stringify(
    DocumentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest$outboundSchema
      .parse(
        documentLookupV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesRetrievableDocumentIdGetRequest,
      ),
  );
}
