// @mistralai/mistralai/src/models/operations/documentfetchv1ragindexesindexindexidschemasschemaschemaidretrievablesget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1943f3a5a249
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest =
  {
    indexId: string;
    schemaId: string;
    /**
     * Only retrieve from this group
     */
    groupId?: string | null | undefined;
  };

/** @internal */
export type DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest$Outbound =
  {
    index_id: string;
    schema_id: string;
    group_id?: string | null | undefined;
  };

/** @internal */
export const DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest$outboundSchema:
  z.ZodType<
    DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest$Outbound,
    DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest
  > = z.object({
    indexId: z.string(),
    schemaId: z.string(),
    groupId: z.nullable(z.string()).optional(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      schemaId: "schema_id",
      groupId: "group_id",
    });
  });

export function documentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequestToJSON(
  documentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest:
    DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest,
): string {
  return JSON.stringify(
    DocumentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest$outboundSchema
      .parse(
        documentFetchV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdRetrievablesGetRequest,
      ),
  );
}
