// @mistralai/mistralai/src/models/operations/getindexdetailsv1ragindexesindexindexiddetailget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d59107500251
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest = {
  indexId: string;
};

/** @internal */
export type GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest$Outbound = {
  index_id: string;
};

/** @internal */
export const GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest$outboundSchema:
  z.ZodType<
    GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest$Outbound,
    GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest
  > = z.object({
    indexId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
    });
  });

export function getIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequestToJSON(
  getIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest:
    GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest,
): string {
  return JSON.stringify(
    GetIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest$outboundSchema
      .parse(getIndexDetailsV1RagIndexesIndexIndexIdDetailGetRequest),
  );
}
