// @mistralai/mistralai/src/models/operations/getindexschemadetailv1ragindexesindexindexidschemasschemaschemaiddetailget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 941a49a146cf
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest =
  {
    indexId: string;
    schemaId: string;
  };

/** @internal */
export type GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest$Outbound =
  {
    index_id: string;
    schema_id: string;
  };

/** @internal */
export const GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest$outboundSchema:
  z.ZodType<
    GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest$Outbound,
    GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest
  > = z.object({
    indexId: z.string(),
    schemaId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      schemaId: "schema_id",
    });
  });

export function getIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequestToJSON(
  getIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest:
    GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest,
): string {
  return JSON.stringify(
    GetIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest$outboundSchema
      .parse(
        getIndexSchemaDetailV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdDetailGetRequest,
      ),
  );
}
