// @mistralai/mistralai/src/models/operations/getindexschemafilev1ragindexesindexindexidschemasschemaschemaidfileget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e3c91593c56f
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest =
  {
    indexId: string;
    schemaId: string;
  };

/** @internal */
export type GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest$Outbound =
  {
    index_id: string;
    schema_id: string;
  };

/** @internal */
export const GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest$outboundSchema:
  z.ZodType<
    GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest$Outbound,
    GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest
  > = z.object({
    indexId: z.string(),
    schemaId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      schemaId: "schema_id",
    });
  });

export function getIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequestToJSON(
  getIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest:
    GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest,
): string {
  return JSON.stringify(
    GetIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest$outboundSchema
      .parse(
        getIndexSchemaFileV1RagIndexesIndexIndexIdSchemasSchemaSchemaIdFileGetRequest,
      ),
  );
}
