// @mistralai/mistralai/src/models/operations/getlogfieldoptionsv1observabilitylogsfieldsfieldnameoptionsget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 48528667de4a
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest =
  {
    fieldName: string;
    from?: Date | null | undefined;
    to?: Date | null | undefined;
  };

/** @internal */
export type GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest$Outbound =
  {
    field_name: string;
    from?: string | null | undefined;
    to?: string | null | undefined;
  };

/** @internal */
export const GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest$outboundSchema:
  z.ZodType<
    GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest$Outbound,
    GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest
  > = z.object({
    fieldName: z.string(),
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
  }).transform((v) => {
    return remap$(v, {
      fieldName: "field_name",
    });
  });

export function getLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequestToJSON(
  getLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest:
    GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest,
): string {
  return JSON.stringify(
    GetLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest$outboundSchema
      .parse(
        getLogFieldOptionsV1ObservabilityLogsFieldsFieldNameOptionsGetRequest,
      ),
  );
}
