// @mistralai/mistralai/src/models/operations/gettracespansv1observabilitytracestraceidspansget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3383d23e2c2e
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest = {
  traceId: string;
  from?: Date | null | undefined;
  to?: Date | null | undefined;
  pageSize?: number | undefined;
  cursor?: string | null | undefined;
};

/** @internal */
export type GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest$Outbound =
  {
    trace_id: string;
    from?: string | null | undefined;
    to?: string | null | undefined;
    page_size: number;
    cursor?: string | null | undefined;
  };

/** @internal */
export const GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest$outboundSchema:
  z.ZodType<
    GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest$Outbound,
    GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest
  > = z.object({
    traceId: z.string(),
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    pageSize: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
  }).transform((v) => {
    return remap$(v, {
      traceId: "trace_id",
      pageSize: "page_size",
    });
  });

export function getTraceSpansV1ObservabilityTracesTraceIdSpansGetRequestToJSON(
  getTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest:
    GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest,
): string {
  return JSON.stringify(
    GetTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest$outboundSchema
      .parse(getTraceSpansV1ObservabilityTracesTraceIdSpansGetRequest),
  );
}
