// @mistralai/mistralai/src/models/operations/gettracebyidv1observabilitytracestraceidget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c7e01490bd18
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetTraceByIdV1ObservabilityTracesTraceIdGetRequest = {
  traceId: string;
};

/** @internal */
export type GetTraceByIdV1ObservabilityTracesTraceIdGetRequest$Outbound = {
  trace_id: string;
};

/** @internal */
export const GetTraceByIdV1ObservabilityTracesTraceIdGetRequest$outboundSchema:
  z.ZodType<
    GetTraceByIdV1ObservabilityTracesTraceIdGetRequest$Outbound,
    GetTraceByIdV1ObservabilityTracesTraceIdGetRequest
  > = z.object({
    traceId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      traceId: "trace_id",
    });
  });

export function getTraceByIdV1ObservabilityTracesTraceIdGetRequestToJSON(
  getTraceByIdV1ObservabilityTracesTraceIdGetRequest:
    GetTraceByIdV1ObservabilityTracesTraceIdGetRequest,
): string {
  return JSON.stringify(
    GetTraceByIdV1ObservabilityTracesTraceIdGetRequest$outboundSchema.parse(
      getTraceByIdV1ObservabilityTracesTraceIdGetRequest,
    ),
  );
}
