// @mistralai/mistralai/src/models/operations/searchspanevaluationsv1observabilityspansevaluationssearchpost.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5dd192d44063
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest =
  {
    from?: Date | null | undefined;
    to?: Date | null | undefined;
    pageSize?: number | undefined;
    cursor?: string | null | undefined;
    spanEvaluationsRequest: components.SpanEvaluationsRequest;
  };

/** @internal */
export type SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$Outbound =
  {
    from?: string | null | undefined;
    to?: string | null | undefined;
    page_size: number;
    cursor?: string | null | undefined;
    SpanEvaluationsRequest: components.SpanEvaluationsRequest$Outbound;
  };

/** @internal */
export const SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$outboundSchema:
  z.ZodType<
    SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$Outbound,
    SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest
  > = z.object({
    from: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    to: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    pageSize: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
    spanEvaluationsRequest: components.SpanEvaluationsRequest$outboundSchema,
  }).transform((v) => {
    return remap$(v, {
      pageSize: "page_size",
      spanEvaluationsRequest: "SpanEvaluationsRequest",
    });
  });

export function searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequestToJSON(
  searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest:
    SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest,
): string {
  return JSON.stringify(
    SearchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest$outboundSchema
      .parse(
        searchSpanEvaluationsV1ObservabilitySpansEvaluationsSearchPostRequest,
      ),
  );
}
