// @mistralai/mistralai/src/models/operations/setindexsummaryv1ragindexesindexindexidsummaryfieldput.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f71d3eca2a6a
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest = {
  indexId: string;
  updateIndexSummaryRequestSummary: components.UpdateIndexSummaryRequestSummary;
};

/** @internal */
export type SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest$Outbound =
  {
    index_id: string;
    UpdateIndexSummaryRequestSummary:
      components.UpdateIndexSummaryRequestSummary$Outbound;
  };

/** @internal */
export const SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest$outboundSchema:
  z.ZodType<
    SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest$Outbound,
    SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest
  > = z.object({
    indexId: z.string(),
    updateIndexSummaryRequestSummary:
      components.UpdateIndexSummaryRequestSummary$outboundSchema,
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      updateIndexSummaryRequestSummary: "UpdateIndexSummaryRequestSummary",
    });
  });

export function setIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequestToJSON(
  setIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest:
    SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest,
): string {
  return JSON.stringify(
    SetIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest$outboundSchema
      .parse(setIndexSummaryV1RagIndexesIndexIndexIdSummaryFieldPutRequest),
  );
}
