// @mistralai/mistralai/src/models/operations/unregistersearchindexv1ragindexesindexindexiddelete.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: cfd3a1f75216
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest = {
  indexId: string;
};

/** @internal */
export type UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest$Outbound =
  {
    index_id: string;
  };

/** @internal */
export const UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest$outboundSchema:
  z.ZodType<
    UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest$Outbound,
    UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest
  > = z.object({
    indexId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
    });
  });

export function unregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequestToJSON(
  unregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest:
    UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest,
): string {
  return JSON.stringify(
    UnregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest$outboundSchema
      .parse(unregisterSearchIndexV1RagIndexesIndexIndexIdDeleteRequest),
  );
}
