// @mistralai/mistralai/src/models/operations/updateindexmetricsv1ragindexesindexindexidmetricsput.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0419a3478b8c
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type MetricsData =
  | components.UpdateIndexMetricsRequestIndexMetricsOnline
  | components.UpdateIndexMetricsRequestIndexMetricsOffline;

export type UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest = {
  indexId: string;
  requestBody:
    | components.UpdateIndexMetricsRequestIndexMetricsOnline
    | components.UpdateIndexMetricsRequestIndexMetricsOffline;
};

/** @internal */
export type MetricsData$Outbound =
  | components.UpdateIndexMetricsRequestIndexMetricsOnline$Outbound
  | components.UpdateIndexMetricsRequestIndexMetricsOffline$Outbound;

/** @internal */
export const MetricsData$outboundSchema: z.ZodType<
  MetricsData$Outbound,
  MetricsData
> = z.union([
  components.UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema,
  components.UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema,
]);

export function metricsDataToJSON(metricsData: MetricsData): string {
  return JSON.stringify(MetricsData$outboundSchema.parse(metricsData));
}

/** @internal */
export type UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$Outbound =
  {
    index_id: string;
    RequestBody:
      | components.UpdateIndexMetricsRequestIndexMetricsOnline$Outbound
      | components.UpdateIndexMetricsRequestIndexMetricsOffline$Outbound;
  };

/** @internal */
export const UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$outboundSchema:
  z.ZodType<
    UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$Outbound,
    UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest
  > = z.object({
    indexId: z.string(),
    requestBody: z.union([
      components.UpdateIndexMetricsRequestIndexMetricsOnline$outboundSchema,
      components.UpdateIndexMetricsRequestIndexMetricsOffline$outboundSchema,
    ]),
  }).transform((v) => {
    return remap$(v, {
      indexId: "index_id",
      requestBody: "RequestBody",
    });
  });

export function updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequestToJSON(
  updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest:
    UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest,
): string {
  return JSON.stringify(
    UpdateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest$outboundSchema
      .parse(updateIndexMetricsV1RagIndexesIndexIndexIdMetricsPutRequest),
  );
}
